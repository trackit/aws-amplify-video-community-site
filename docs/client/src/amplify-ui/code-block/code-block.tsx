import {Component, Host, h, Prop, State} from "@stencil/core";
import {
  codeBlockStyle,
  lineCountStyle,
  noLineNumbersStyle,
  copyButtonStyle,
  slotContainerStyle,
} from "./code-block.style";
import copy from "copy-to-clipboard";

const COPY = "copy";
const COPIED = "copied";
const FAILED = "failed to copy";
const CONSOLE = "console";

let alwaysRecompute = 0;

@Component({tag: "amplify-code-block", shadow: false})
export class AmplifyCodeBlock {
  /** the number of lines of the code block */
  @Prop() readonly lineCount: number;
  /** what language are we displaying */
  @Prop() readonly language: string;

  @State() copyMessage: typeof COPY | typeof COPIED | typeof FAILED = "copy";

  element?: HTMLDivElement;

  setElement = (ref: HTMLDivElement | undefined) => {
    if (ref !== null) {
      this.element = ref;
    }
  };

  copyToClipboard = () => {
    if (this.element && this.element.textContent) {
      copy(this.element.textContent);
      this.copyMessage = COPIED;
    } else {
      this.copyMessage = FAILED;
    }
    setTimeout(() => {
      this.copyMessage = COPY;
    }, 1500);
  };

  copyButton = () => {
    if (this.language !== CONSOLE) {
      return (
        <button class={copyButtonStyle} onClick={this.copyToClipboard}>
          <span>{this.copyMessage}</span>
        </button>
      );
    }
  };

  lineNumbers = () => {
    if (this.lineCount > 1 && this.language !== CONSOLE) {
      return (
        <div class={lineCountStyle}>
          <div>
            {new Array(this.lineCount).fill(null).map((_, i) => (
              <span>{String(i + 1)}</span>
            ))}
          </div>
        </div>
      );
    }
  };

  render() {
    return (
      <Host
        class={{
          [codeBlockStyle]: true,
          [noLineNumbersStyle]:
            this.lineCount === 1 || this.language === CONSOLE,
        }}
      >
        {this.lineNumbers()}
        <div
          key={String(alwaysRecompute++)}
          ref={this.setElement}
          class={slotContainerStyle}
        >
          <slot name="content" />
        </div>
        {this.copyButton()}
      </Host>
    );
  }
}
