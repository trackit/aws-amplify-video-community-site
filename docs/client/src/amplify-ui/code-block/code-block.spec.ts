import {AmplifyCodeBlock} from "./code-block";
import {newSpecPage} from "@stencil/core/testing";

describe("amplify-code-block", () => {
  describe("Render logic", () => {
    it("should render", async () => {
      expect(
        (
          await newSpecPage({
            components: [AmplifyCodeBlock],
            html: `<amplify-code-block />`,
          })
        ).root,
      ).toMatchSnapshot();
    });
  });

  describe("Copy button", () => {
    it("is visible for languages other than console", async () => {
      expect(
        (
          await newSpecPage({
            components: [AmplifyCodeBlock],
            html: `<amplify-code-block language="javascript">
                     const lol = 'i dunno';
                     alert(lol);
                   </amplify-code-block>`,
          })
        ).root,
      ).toMatchSnapshot();
    });

    it("is not visible if the language is set to console", async () => {
      expect(
        (
          await newSpecPage({
            components: [AmplifyCodeBlock],
            html: `<amplify-code-block language="console" />`,
          })
        ).root,
      ).toMatchSnapshot();
    });
  });

  describe("Line numbers", () => {
    it("is visible if the code block has more than one line", async () => {
      expect(
        (
          await newSpecPage({
            components: [AmplifyCodeBlock],
            html: `<amplify-code-block language="javascript">
                     const lol = 'i dunno';
                     alert(lol);
                   </amplify-code-block>`,
          })
        ).root,
      ).toMatchSnapshot();
    });

    it("is not visible if the code block only has a single line", async () => {
      expect(
        (
          await newSpecPage({
            components: [AmplifyCodeBlock],
            html: `<amplify-code-block language="console">
                   line 1
                   line 2
                   line 3
                   </amplify-code-block>`,
          })
        ).root,
      ).toMatchSnapshot();
    });

    it("is not visible if the language is set to console", async () => {
      expect(
        (
          await newSpecPage({
            components: [AmplifyCodeBlock],
            html: `<amplify-code-block language="console">
                   line 1
                   line 2
                   line 3
                   </amplify-code-block>`,
          })
        ).root,
      ).toMatchSnapshot();
    });
  });
});
