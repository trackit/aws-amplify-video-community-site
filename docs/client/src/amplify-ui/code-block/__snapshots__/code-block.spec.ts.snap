// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`amplify-code-block Copy button is not visible if the language is set to console 1`] = `
<amplify-code-block class="css-1lj8awz css-75wbz6" language="console">
  <!---->
  <div class="css-khgmir"></div>
</amplify-code-block>
`;

exports[`amplify-code-block Copy button is visible for languages other than console 1`] = `
<amplify-code-block class="css-1lj8awz" language="javascript">
  <!---->
  const lol = 'i dunno'; alert(lol);
  <div class="css-khgmir"></div>
  <button class="css-16vr62x">
    <span>
      copy
    </span>
  </button>
</amplify-code-block>
`;

exports[`amplify-code-block Line numbers is not visible if the code block only has a single line 1`] = `
<amplify-code-block class="css-1lj8awz css-75wbz6" language="console">
  <!---->
  line 1 line 2 line 3
  <div class="css-khgmir"></div>
</amplify-code-block>
`;

exports[`amplify-code-block Line numbers is not visible if the language is set to console 1`] = `
<amplify-code-block class="css-1lj8awz css-75wbz6" language="console">
  <!---->
  line 1 line 2 line 3
  <div class="css-khgmir"></div>
</amplify-code-block>
`;

exports[`amplify-code-block Line numbers is visible if the code block has more than one line 1`] = `
<amplify-code-block class="css-1lj8awz" language="javascript">
  <!---->
  const lol = 'i dunno'; alert(lol);
  <div class="css-khgmir"></div>
  <button class="css-16vr62x">
    <span>
      copy
    </span>
  </button>
</amplify-code-block>
`;

exports[`amplify-code-block Render logic should render 1`] = `
<amplify-code-block class="css-1lj8awz">
  <!---->
  <div class="css-khgmir"></div>
  <button class="css-16vr62x">
    <span>
      copy
    </span>
  </button>
</amplify-code-block>
`;
